lib/reduceBy/reduceBy.test.js

```diff
@@ -1,21 +1,50 @@
+import always from '../always'
+import cond from '../cond'
+import lt from '../lt'
 import reduceBy from '.'
+import T from '../T'
 
-test('groups up reduced values by callback from 3rd argument', () => {
-  var reduceToNamesBy = reduceBy((acc, student) => acc.concat(student.name), [])
-  var namesByGrade = reduceToNamesBy(function(student) {
-    var score = student.score
-    return score < 65
-      ? 'F'
-      : score < 70 ? 'D' : score < 80 ? 'C' : score < 90 ? 'B' : 'A'
-  })
-  var students = [
+const lt_ = than => what => lt(what, than)
+
+it('groups up reduced values by callback from 3rd argument', () => {
+  const reduceToNamesBy = reduceBy((names, { name }) => {
+    names.push(name)
+    return names
+  }, [])
+  const namesByGrade = reduceToNamesBy(({ score }) => cond([
+    [lt_(65), always('E')],
+    [lt_(70), always('D')],
+    [lt_(80), always('C')],
+    [lt_(90), always('B')],
+    [T, always('A')]
+  ])(score))
+  const students = [
     { name: 'Lucy', score: 92 },
     { name: 'Drew', score: 85 },
     { name: 'Bart', score: 62 }
   ]
   expect(namesByGrade(students)).toEqual({
     A: ['Lucy'],
     B: ['Drew'],
-    F: ['Bart']
+    E: ['Bart']
   })
 })
+
+it('shallowly clones initial value for groups', () => {
```

Comment:
It's hard to understand because of too many variables. I see it smth like this:

```js
const initialValue = [{ deepTest: true }]
  const students = [
    { name: 'Lucy', score: 92 },
    { name: 'Drew', score: 85 },
    { name: 'Bart', score: 62 }
  ]

  // transform data with reduceBy
  const result = reduceBy(
    (names, { name }) => [...names, name],
    initialValue,
    ({ name }) => name[0].toUpperCase(),
    students
  )

  const values = Object.values(result)
  
  // mutate initialData value
  initialValue[0].deepTest = false

  // check mutated data
  for (let i = 0; i < values.length; i++) {
    expect(values[i][0].deepText).toBeFalsy()
  }
```

It's just proposal, you could not use it :)

Reply:
Thank you
Done